docs/standard/native-interop/comwrappers-source-generation.md

```diff
@@ -0,0 +1,196 @@
+    {
+        // Do things
+    }
+}
+```
+
+At compile time, the generator will create an implementation of the ComWrappers API, and you can use the <xref:System.Runtime.InteropServices.Marshalling.StrategyBasedComWrappers> type or a custom derived type to consume or expose the COM interface.
+
+```csharp
+[LibraryImport("MyComObjectProvider.dll")]
+static nint GetPointerToComInterface();
+
+[LibraryImport("MyComObjectProvider.dll")]
+static void GivePointerToComInterface(nint comObject);
+
+// Use the system to create a Runtime Callable Wrapper to use in managed code
+ComWrappers cw = new StrategyBasedComWrappers();
+nint ptr = GetPointerToComInterface();
+IFoo foo = (IFoo)cw.GetOrCreateObjectForComInterface(ptr);
+foo.Method(0);
+...
+// Use the system to create a COM Callable Wrapper to pass to unmanaged code
+ComWrappers cw = new StrategyBasedComWrappers();
+Foo foo = new();
+nint ptr = cw.GetOrCreateComInterfaceForObject(foo);
+GivePointerToComInterface(ptr);
+```
+
+## Customizing marshalling
+
+The COM interface generator respects the <xref:System.Runtime.InteropServices.MarshalAsAttribute> and <xref:System.Runtime.InteropServices.Marshalling.MarshalUsingAttribute> attributes to customize marshalling of parameters. For more information, see [Customize Parameter Marshalling](./customize-parameter-marshalling.md) If an interface has `string` parameters or return types, the <xref:System.Runtime.InteropServices.Marshalling.GeneratedComInterfaceAttribute.StringMarshalling?displayProperty=nameWithType> and <xref:System.Runtime.InteropServices.Marshalling.GeneratedComInterfaceAttribute.StringMarshallingCustomType?displayProperty=nameWithType> properties will apply to all parameters and return types of `string` in the interface if they don't have other marshalling attributes.
+
+## Implicit HRESULTs and PreserveSig
+
+COM methods in C# have a different signature than the native methods. Standard COM has a return type of `HRESULT`, a 4 byte integer type representing error and success states. This `HRESULT` return value is hidden by default in the C# signature and converted to an exception when an error value is returned. The last "out" parameter of the native COM signature may optionally be converted into the return in the C# signature.
+
+For example, the following snippets show C# method signatures and the corresponding native signature the generator infers.
+
+```csharp
+void Method1(int i);
+
+int Method2(float i);
+```
+
+```c
+HRESULT Method1(int i);
+
+HRESULT Method2(float i, _Out_ int* returnValue);
+```
+
+If you want to handle the HRESULT yourself, you can use the <xref:System.Runtime.InteropServices.PreserveSigAttribute> on the method to indicate the generator should not do this transformation. Below are snippets that demonstrate what native signature the generator expects when `[PreserveSig]` is applied.
+
+```csharp
+[PreserveSig]
+void Method1(int i);
+
+[PreserveSig]
+int Method2(float i);
+```
+
+```c
+int Method1(int i);
```

Comment:
```suggestion
void Method1(int i);
```

Reply:
Changed the managed signature to return an int